# Data & Web Mining
## Python crash course
## Numpy and Pandas


#### Prof. Claudio Lucchese

## NumPy 

http://www.numpy.org/

**NumPy** stands for **Numerical Python**.

NumPy is a fundamental package for efficient scientific and numerical computing.

It provides:
- efficient methods for managing arrays and matrices, and operations on them
- several mathematical functions (variance, standard deviation, cumulative sum, ...)
- other: fitting a polynomial, finding the minimum of a function, Fast Fourier Transform, etc.


## Matrix representation and operations

In [1]:
# transform a nested list into a 2D matrix

import numpy as np
a = np.array( [ [1.,2.,3.],
                [4.,5.,6.] ] )

print ("Matrix a:")
print (a)

Matrix a:
[[1. 2. 3.]
 [4. 5. 6.]]


In [2]:
# Element-wise operations
b = a + 2
print ("Matrix b=a+2:")
print (b)
print()

c = a * 7
print ("Matrix c=a*7:")
print (c)
print()

d = b - a
print ("Matrix d=b-a:")
print (d)
print()

e = c / a
print ("Matrix e=c/a:")
print (e)

Matrix b=a+2:
[[3. 4. 5.]
 [6. 7. 8.]]

Matrix c=a*7:
[[ 7. 14. 21.]
 [28. 35. 42.]]

Matrix d=b-a:
[[2. 2. 2.]
 [2. 2. 2.]]

Matrix e=c/a:
[[7. 7. 7.]
 [7. 7. 7.]]


In [3]:
# Matrix transpose and multiplication
z = np.matmul(a, b.T)
print (z)

[[ 26.  44.]
 [ 62. 107.]]


## Mathematical functions

See https://docs.scipy.org/doc/numpy/reference/ufuncs.html for more information.

Let's test `sqrt()` (square root) and `maximum()` (element-wise maximum of two arrays).

In [4]:
a = np.array( [ [1.,2.,3.],[4.,5.,6.] ] )
b = np.array( [ [3.,3.,7.],[1.,1.,2.] ] )

print ("exp function")
print ( np.exp(a))
print ()

print ("element-wise maximum")
print ( np.maximum(a,b) )

exp function
[[  2.71828183   7.3890561   20.08553692]
 [ 54.59815003 148.4131591  403.42879349]]

element-wise maximum
[[3. 3. 7.]
 [4. 5. 6.]]


## Aggregation and Statistical methods

Typical `sum`, `mean`, `var`, are available ...

In case of matrices, it is possible to specify the **direction** of the operation.
 - see https://docs.scipy.org/doc/numpy/reference/generated/numpy.mean.html#numpy.mean

In [5]:
m = np.array( [ [1.,2.,3.],
                [4.,5.,6.] ] )

print (m)
print ()

print ("No direction/axis")
print ( np.mean(m) )
print ()

print ("mean over axis 0 (across rows)")
print ( np.mean(m, axis=0) )
print ()

print ("mean over axis 1 (across cols)")
print ( np.mean(m, axis=1) )
print ()


[[1. 2. 3.]
 [4. 5. 6.]]

No direction/axis
3.5

mean over axis 0 (across rows)
[2.5 3.5 4.5]

mean over axis 1 (across cols)
[2. 5.]



## Indexing and Views

It is possible to access elements of matrices/array in a similar way to Python lists.

Slices of a matrix are implemented as **views**, **not copies**, on the original data, sharing the same memory.

In [6]:
m = np.array( [ [1.,2.,3.],
                [4.,5.,6.] ] )

print (m[0,2])
print ()

print ("view/indexing")
view = m[:,1]  # all rows, column with index 1
print ( view )
print ()

print ("modified view and original data")
view[:] = 33
print ( view )
print ()
print ( m )    # also m is modified

3.0

view/indexing
[2. 5.]

modified view and original data
[33. 33.]

[[ 1. 33.  3.]
 [ 4. 33.  6.]]


## Sorting and Fancy Indexing

Similarly to python, the sort method can be used to sort an array or to get a sorted copy.

There is no `key` parameter.

See https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.sort.html.

In [7]:
data = np.array([1,-2,3,-4,5])

s = np.sort(data)
print ( "sorted copy    ", s )
print ( "original data  ", data)
print()

data.sort()
print ( "after data.sort()" )
print ( "original data  ", data)

sorted copy     [-4 -2  1  3  5]
original data   [ 1 -2  3 -4  5]

after data.sort()
original data   [-4 -2  1  3  5]


Also sorting may have a direction.

In [8]:
m = np.array( [ [3., 5.,2.],[6., 1.,5.] ] )

print (m)

print ("sort over axis 0 (across rows)")
print ( np.sort(m, axis=0) )
print()

print ("sort over axis 1 (across cols)")
print ( np.sort(m, axis=1) )
print()

print ("flatten and then sort")
print ( np.sort(m, axis=None) )

[[3. 5. 2.]
 [6. 1. 5.]]
sort over axis 0 (across rows)
[[3. 1. 2.]
 [6. 5. 5.]]

sort over axis 1 (across cols)
[[2. 3. 5.]
 [1. 5. 6.]]

flatten and then sort
[1. 2. 3. 5. 5. 6.]


A useful method is `argsort`, which returns the positions of the elements in sorted order, withouth modifying the original array.

The output of `argsort`, can be used in conjuction with fancy indexing.

Example: sort by income and print the corresponding name.

In [9]:
names = np.array([  'Mark', 'Joe', 'Will', 'Bob', 'Jane', 'Carol', 'Donald'])
salaries = np.array([2000,   1200,  3000,  2100,   1580,   1700,    900])

sorted_pos = np.argsort(salaries)
print ("sorted positions", sorted_pos)

print ( "salaries",  salaries[sorted_pos] ) # This is called fancy indexing!
print ( "ages",  names[sorted_pos] ) # This is called fancy indexing!

sorted positions [6 1 4 5 0 3 2]
salaries [ 900 1200 1580 1700 2000 2100 3000]
ages ['Donald' 'Joe' 'Jane' 'Carol' 'Mark' 'Bob' 'Will']


## Optimization

Find the root of a function given it first derivative.

See https://docs.scipy.org/doc/scipy/reference/optimize.html

In [10]:
from scipy import optimize
def f(x):
    return (x**3 - 1)  # only one real root at x = 1

def fprime(x):
    return 3*x**2

sol = optimize.root_scalar( f,                # function
                            x0=0.2,           # initial guess
                            fprime=fprime,    # first derivative
                            method='newton')  # optimization method

print ("The root of the function is:", sol.root)

The root of the function is: 1.0


# More Numpy ?

NumPy deserves your interest, especially for implementing numerical algorithms, matrix-based operations, and to exploit its great algorithms (see also scipy).

From a data perspective, it provides a low level access. We will see the Pandas library, which is more data-oriented and it shares several commonalities with NumPy.

## Pandas: Python Data Analysis Library

See: https://pandas.pydata.org/

Pandas is an open source providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.

## Let's read a dataset with one line of code!

We can read an excel file!

Data available at http://tennis-data.co.uk/alldata.php.

See: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html

Check also the `sheet_name` parameter.

In [11]:
import pandas as pd

dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'

df = pd.read_excel(dataset_file) 
df.head(5) # see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html

ATP  Location              Tournament       Date  Series    Court Surface  \
0    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
1    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
2    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
3    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
4    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   

       Round  Best of         Winner  ... Lsets    Comment  B365W  B365L  \
0  1st Round        3    Dimitrov G.  ...   0.0  Completed   1.36   3.00   
1  1st Round        3      Raonic M.  ...   0.0  Completed   1.18   4.50   
2  1st Round        3  Kecmanovic M.  ...   0.0  Completed   1.57   2.25   
3  1st Round        3     Millman J.  ...   1.0  Completed   1.40   2.75   
4  1st Round        3    Uchiyama Y.  ...   0.0  Completed   2.62   1.44   

    PSW   PSL  MaxW  MaxL  AvgW  AvgL  
0  1.36  3.37  1.42  3.60  1.35  3.18  
1  1.23  4.68  1.27  4.84  1.22  4.26  
2  1.67  2.32  1.71  2.40  1.63  2.28  
3  1.41  3.13  1.45  3.20  1.40  2.95  
4  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 36 columns]

In [12]:
print (type(df))
# see https://pandas.pydata.org/pandas-docs/stable/reference/frame.html

<class 'pandas.core.frame.DataFrame'>


## Dataframe

- **DataFrame** is a 2-dimensional table of data and contains an ordered collection of columns, each of which can be a different value type (numeric, string, boolean, etc.).

See: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html

**Features of DataFrame:**
- Mutable
- Potentially columns are of different types
- Labeled axes (rows and columns)
- Can Perform Arithmetic operations on rows and columns

A DataFrame can be seen as a dictionary of columns (indeed, pandas Series), all sharing the same index. 

We can ask the number of rows and columns.
We can access index and columns labels with attributes `index`and `column`.

In [13]:
print("df.shape:", df.shape)
print("df.index:", df.index)
print("df.columns:", df.columns)

df.shape: (2610, 36)
df.index: RangeIndex(start=0, stop=2610, step=1)
df.columns: Index(['ATP', 'Location', 'Tournament', 'Date', 'Series', 'Court', 'Surface',
       'Round', 'Best of', 'Winner', 'Loser', 'WRank', 'LRank', 'WPts', 'LPts',
       'W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'W4', 'L4', 'W5', 'L5', 'Wsets',
       'Lsets', 'Comment', 'B365W', 'B365L', 'PSW', 'PSL', 'MaxW', 'MaxL',
       'AvgW', 'AvgL'],
      dtype='object')


## Question: What is the number of matches?


In [14]:
import pandas as pd

dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'
df = pd.read_excel(dataset_file) 

print ("Dataframe shape is:", df.shape)
num_matches, num_columns = df.shape
print ("The number of matches is:", num_matches)

Dataframe shape is: (2610, 36)
The number of matches is: 2610


### Names

Both index and columns have an attribute `name` to specify their names.


In [15]:
df.index.name   = "Match ID"
df.columns.name = "Features"

df.head()

Features  ATP  Location              Tournament       Date  Series    Court  \
Match ID                                                                      
0           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
1           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
2           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
3           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
4           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   

Features Surface      Round  Best of         Winner  ... Lsets    Comment  \
Match ID                                             ...                    
0           Hard  1st Round        3    Dimitrov G.  ...   0.0  Completed   
1           Hard  1st Round        3      Raonic M.  ...   0.0  Completed   
2           Hard  1st Round        3  Kecmanovic M.  ...   0.0  Completed   
3           Hard  1st Round        3     Millman J.  ...   1.0  Completed   
4           Hard  1st Round        3    Uchiyama Y.  ...   0.0  Completed   

Features  B365W  B365L   PSW   PSL  MaxW  MaxL  AvgW  AvgL  
Match ID                                                    
0          1.36   3.00  1.36  3.37  1.42  3.60  1.35  3.18  
1          1.18   4.50  1.23  4.68  1.27  4.84  1.22  4.26  
2          1.57   2.25  1.67  2.32  1.71  2.40  1.63  2.28  
3          1.40   2.75  1.41  3.13  1.45  3.20  1.40  2.95  
4          2.62   1.44  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 36 columns]

## Statistical summary

We can get a statistical summary for numerical columns.

In [16]:
df.describe()

Features          ATP      Best of        WRank        LRank          WPts  \
count     2610.000000  2610.000000  2606.000000  2597.000000   2607.000000   
mean        33.218774     3.388506    57.129317    77.825568   1782.094745   
std         18.226778     0.791400    57.173452    80.276616   2055.165260   
min          1.000000     3.000000     1.000000     1.000000     17.000000   
25%         19.000000     3.000000    19.000000    34.000000    716.000000   
50%         32.000000     3.000000    45.000000    60.000000   1022.000000   
75%         50.000000     3.000000    76.000000    95.000000   1825.000000   
max         66.000000     5.000000   503.000000  1491.000000  12415.000000   

Features          LPts           W1           L1           W2           L2  \
count      2597.000000  2589.000000  2589.000000  2576.000000  2576.000000   
mean       1156.386985     5.826574     4.202008     5.834239     4.036879   
std        1194.664111     1.201032     1.822849     1.213551     1.830008   
min           2.000000     0.000000     0.000000     0.000000     0.000000   
25%         598.000000     6.000000     3.000000     6.000000     3.000000   
50%         875.000000     6.000000     4.000000     6.000000     4.000000   
75%        1246.000000     6.000000     6.000000     6.000000     6.000000   
max       12355.000000     7.000000     7.000000     7.000000     7.000000   

Features  ...        Wsets        Lsets        B365W        B365L  \
count     ...  2589.000000  2589.000000  2597.000000  2597.000000   
mean      ...     2.167246     0.431054     1.844926     3.159365   
std       ...     0.435385     0.565270     0.893305     3.067155   
min       ...     0.000000     0.000000     1.002000     1.070000   
25%       ...     2.000000     0.000000     1.300000     1.660000   
50%       ...     2.000000     0.000000     1.570000     2.300000   
75%       ...     2.000000     1.000000     2.100000     3.400000   
max       ...     3.000000     2.000000     9.000000    41.000000   

Features          PSW          PSL         MaxW         MaxL         AvgW  \
count     2599.000000  2599.000000  2609.000000  2609.000000  2609.000000   
mean         1.933342     3.474467     1.998605     3.678862     1.867348   
std          1.027109     3.597585     1.096063     4.221048     0.910771   
min          1.005000     1.070000     1.010000     1.080000     1.010000   
25%          1.330000     1.760000     1.360000     1.820000     1.310000   
50%          1.640000     2.440000     1.670000     2.510000     1.600000   
75%          2.185000     3.625000     2.270000     3.780000     2.120000   
max         11.730000    37.800000    12.220000    67.000000     9.640000   

Features         AvgL  
count     2609.000000  
mean         3.139747  
std          2.668966  
min          1.060000  
25%          1.720000  
50%          2.340000  
75%          3.430000  
max         28.490000  

[8 rows x 26 columns]

## Data Types

A specific data type is used to store and manage the information in the dataframe. This is important to understand which operations can be performed on the different columns.

Note that also the non-null values are reported. It is not uncommon to have missing values in our dataset.

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2610 entries, 0 to 2609
Data columns (total 36 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   ATP         2610 non-null   int64         
 1   Location    2610 non-null   object        
 2   Tournament  2610 non-null   object        
 3   Date        2610 non-null   datetime64[ns]
 4   Series      2610 non-null   object        
 5   Court       2610 non-null   object        
 6   Surface     2610 non-null   object        
 7   Round       2610 non-null   object        
 8   Best of     2610 non-null   int64         
 9   Winner      2610 non-null   object        
 10  Loser       2610 non-null   object        
 11  WRank       2606 non-null   float64       
 12  LRank       2597 non-null   float64       
 13  WPts        2607 non-null   float64       
 14  LPts        2597 non-null   float64       
 15  W1          2589 non-null   float64       
 16  L1          2589 non-nul

## Column selection, addition, and deletion

See: https://pandas.pydata.org/pandas-docs/stable/indexing.html.

Pandas provides several (sometimes confusing) ways to access the columns of a data frame.

#### Example (Selection)
We can select a column by specifying its name. 

This operation gives us Pandas Series.

Note that the index is preserved.

In [18]:
a = df["Winner"] 

print("a:\n", a)
print("Type: ", type(a))
print()

# Equivalent to

a = df.Winner
print("a:\n", a)
print("Type: ", type(a))

a:
 Match ID
0         Dimitrov G.
1           Raonic M.
2       Kecmanovic M.
3          Millman J.
4         Uchiyama Y.
            ...      
2605        Nadal R. 
2606       Zverev A. 
2607    Tsitsipas S. 
2608        Thiem D. 
2609    Tsitsipas S. 
Name: Winner, Length: 2610, dtype: object
Type:  <class 'pandas.core.series.Series'>

a:
 Match ID
0         Dimitrov G.
1           Raonic M.
2       Kecmanovic M.
3          Millman J.
4         Uchiyama Y.
            ...      
2605        Nadal R. 
2606       Zverev A. 
2607    Tsitsipas S. 
2608        Thiem D. 
2609    Tsitsipas S. 
Name: Winner, Length: 2610, dtype: object
Type:  <class 'pandas.core.series.Series'>


You can also select multiple columns. In this case you need to use a list.

Note that the result is a dataframe.

In [19]:
wl = df[ ["Winner","Loser"] ] 

wl.head()

Features         Winner        Loser
Match ID                            
0           Dimitrov G.  Nishioka Y.
1             Raonic M.    Bedene A.
2         Kecmanovic M.     Mayer L.
3            Millman J.  Sandgren T.
4           Uchiyama Y.   Humbert U.

#### Example (Addition)
We can add a new column by passing a NumPy array (or a list, Pandas Series, and more) or a single number.

In case of a list/array, the length of the array must equal the number of rows (otherwise a ValueError exception is raised)!

In [20]:
print("df.shape:", df.shape)

         # note that len gives us the number of rows
df['meaning-less']   = np.ones( len(df) ) 
df['meaning-less-2'] = 2
df['meaning-less-3'] = df['B365W']*2

df.head()

df.shape: (2610, 36)


Features  ATP  Location              Tournament       Date  Series    Court  \
Match ID                                                                      
0           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
1           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
2           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
3           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
4           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   

Features Surface      Round  Best of         Winner  ... B365L   PSW   PSL  \
Match ID                                             ...                     
0           Hard  1st Round        3    Dimitrov G.  ...  3.00  1.36  3.37   
1           Hard  1st Round        3      Raonic M.  ...  4.50  1.23  4.68   
2           Hard  1st Round        3  Kecmanovic M.  ...  2.25  1.67  2.32   
3           Hard  1st Round        3     Millman J.  ...  2.75  1.41  3.13   
4           Hard  1st Round        3    Uchiyama Y.  ...  1.44  2.73  1.51   

Features  MaxW  MaxL  AvgW  AvgL  meaning-less  meaning-less-2  meaning-less-3  
Match ID                                                                        
0         1.42  3.60  1.35  3.18           1.0               2            2.72  
1         1.27  4.84  1.22  4.26           1.0               2            2.36  
2         1.71  2.40  1.63  2.28           1.0               2            3.14  
3         1.45  3.20  1.40  2.95           1.0               2            2.80  
4         3.26  1.53  2.69  1.47           1.0               2            5.24  

[5 rows x 39 columns]

#### Example (Deletion)

Method `drop()` can be used to remove a column. We need to specify axis=1 (axis=0 are the rows).
The built-in `del` can also be used.

`inplace` argument (default is False)  is common in several Pandas' functions that modify the DataFrame. 
If True, it says that the function has to modify the DataFrame itself instead of returning a new one.

In [21]:
        # drop returns a new dataframe
df = df.drop('meaning-less', axis=1)   
        # unless inplace is set to True
df.drop('meaning-less-2', axis=1, inplace=True)
        # alternative
del df['meaning-less-3']

df.head()

Features  ATP  Location              Tournament       Date  Series    Court  \
Match ID                                                                      
0           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
1           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
2           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
3           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
4           1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   

Features Surface      Round  Best of         Winner  ... Lsets    Comment  \
Match ID                                             ...                    
0           Hard  1st Round        3    Dimitrov G.  ...   0.0  Completed   
1           Hard  1st Round        3      Raonic M.  ...   0.0  Completed   
2           Hard  1st Round        3  Kecmanovic M.  ...   0.0  Completed   
3           Hard  1st Round        3     Millman J.  ...   1.0  Completed   
4           Hard  1st Round        3    Uchiyama Y.  ...   0.0  Completed   

Features  B365W  B365L   PSW   PSL  MaxW  MaxL  AvgW  AvgL  
Match ID                                                    
0          1.36   3.00  1.36  3.37  1.42  3.60  1.35  3.18  
1          1.18   4.50  1.23  4.68  1.27  4.84  1.22  4.26  
2          1.57   2.25  1.67  2.32  1.71  2.40  1.63  2.28  
3          1.40   2.75  1.41  3.13  1.45  3.20  1.40  2.95  
4          2.62   1.44  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 36 columns]

## Index

Index does not need to be in integer, and it does not need to be unique.
We can choose one of the column to be the index with function `set_index()`. 

Note that the old index is lost!

In [22]:
df.set_index("Location", inplace=True) # Alternative to df = df.set_index("Location")

# Note the new name of the index
df.head()

Features  ATP              Tournament       Date  Series    Court Surface  \
Location                                                                    
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   

Features      Round  Best of         Winner        Loser  ...  Lsets  \
Location                                                  ...          
Brisbane  1st Round        3    Dimitrov G.  Nishioka Y.  ...    0.0   
Brisbane  1st Round        3      Raonic M.    Bedene A.  ...    0.0   
Brisbane  1st Round        3  Kecmanovic M.     Mayer L.  ...    0.0   
Brisbane  1st Round        3     Millman J.  Sandgren T.  ...    1.0   
Brisbane  1st Round        3    Uchiyama Y.   Humbert U.  ...    0.0   

Features    Comment  B365W  B365L   PSW   PSL  MaxW  MaxL  AvgW  AvgL  
Location                                                               
Brisbane  Completed   1.36   3.00  1.36  3.37  1.42  3.60  1.35  3.18  
Brisbane  Completed   1.18   4.50  1.23  4.68  1.27  4.84  1.22  4.26  
Brisbane  Completed   1.57   2.25  1.67  2.32  1.71  2.40  1.63  2.28  
Brisbane  Completed   1.40   2.75  1.41  3.13  1.45  3.20  1.40  2.95  
Brisbane  Completed   2.62   1.44  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 35 columns]

## Question: List the tournament names

In [23]:
import pandas as pd

dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'
df = pd.read_excel(dataset_file) 

df['Tournament']

0       Brisbane International
1       Brisbane International
2       Brisbane International
3       Brisbane International
4       Brisbane International
                 ...          
2605               Masters Cup
2606               Masters Cup
2607               Masters Cup
2608               Masters Cup
2609               Masters Cup
Name: Tournament, Length: 2610, dtype: object

In [24]:
set( df['Tournament'] )

{'ABN AMRO World Tennis Tournament',
 'ASB Classic',
 'Abierto Mexicano',
 'Abierto Mexicano Mifel',
 'Antalya Open',
 'Argentina Open',
 'Australian Open',
 'BB&T Atlanta Open',
 'BMW Open',
 'BNP Paribas Masters',
 'BNP Paribas Open',
 'Brasil Open',
 'Brisbane International',
 'Chengdu Open',
 'China Open',
 'Citi Open',
 'Cordoba Open',
 'Croatia Open',
 'Delray Beach Open',
 'Dubai Tennis Championships',
 'Eastbourne International',
 'Erste Bank Open',
 'European Open',
 'French Open',
 'Generali Open',
 'Geneva Open',
 'German Tennis Championships',
 'Grand Prix Hassan II',
 'Hall of Fame Championships',
 'Halle Open',
 'Hungarian Open',
 "Internazionali BNL d'Italia",
 'Kremlin Cup',
 'Lyon Open',
 'Maharashtra Open',
 'Masters Cup',
 'Mercedes Cup',
 'Millennium Estoril Open',
 'Monte Carlo Masters',
 'Mutua Madrid Open',
 'New York Open',
 'Open 13',
 'Open Banco Sabadell ',
 'Open Sud de France',
 'Open de Moselle',
 'Qatar Exxon Mobil Open',
 "Queen's Club Championships",
 '

You can do this because a pandas Series is iterable, and a python set can be build from any iterable.

You can iterate in two ways.

In [25]:
for v in df['Tournament']:
    print (v)

Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon

In [26]:
# similar to enumerate
for index,v in df['Tournament'].iteritems():
    print (index, v)

0 Brisbane International
1 Brisbane International
2 Brisbane International
3 Brisbane International
4 Brisbane International
5 Brisbane International
6 Brisbane International
7 Brisbane International
8 Brisbane International
9 Brisbane International
10 Brisbane International
11 Brisbane International
12 Brisbane International
13 Brisbane International
14 Brisbane International
15 Brisbane International
16 Brisbane International
17 Brisbane International
18 Brisbane International
19 Brisbane International
20 Brisbane International
21 Brisbane International
22 Brisbane International
23 Brisbane International
24 Brisbane International
25 Brisbane International
26 Brisbane International
27 Qatar Exxon Mobil Open
28 Qatar Exxon Mobil Open
29 Qatar Exxon Mobil Open
30 Qatar Exxon Mobil Open
31 Qatar Exxon Mobil Open
32 Qatar Exxon Mobil Open
33 Qatar Exxon Mobil Open
34 Qatar Exxon Mobil Open
35 Qatar Exxon Mobil Open
36 Qatar Exxon Mobil Open
37 Qatar Exxon Mobil Open
38 Qatar Exxon Mobil O

1513 Wimbledon
1514 Wimbledon
1515 Wimbledon
1516 Wimbledon
1517 Wimbledon
1518 Wimbledon
1519 Wimbledon
1520 Wimbledon
1521 Wimbledon
1522 Wimbledon
1523 Wimbledon
1524 Wimbledon
1525 Wimbledon
1526 Wimbledon
1527 Wimbledon
1528 Wimbledon
1529 Wimbledon
1530 Wimbledon
1531 Wimbledon
1532 Wimbledon
1533 Wimbledon
1534 Wimbledon
1535 Wimbledon
1536 Wimbledon
1537 Wimbledon
1538 Wimbledon
1539 Wimbledon
1540 Wimbledon
1541 Wimbledon
1542 Wimbledon
1543 Wimbledon
1544 Wimbledon
1545 Wimbledon
1546 Wimbledon
1547 Wimbledon
1548 Wimbledon
1549 Wimbledon
1550 Wimbledon
1551 Wimbledon
1552 Wimbledon
1553 Wimbledon
1554 Wimbledon
1555 Wimbledon
1556 Wimbledon
1557 Wimbledon
1558 Wimbledon
1559 Wimbledon
1560 Wimbledon
1561 Wimbledon
1562 Wimbledon
1563 Wimbledon
1564 Wimbledon
1565 Wimbledon
1566 Wimbledon
1567 Wimbledon
1568 Wimbledon
1569 Wimbledon
1570 Wimbledon
1571 Wimbledon
1572 Wimbledon
1573 Wimbledon
1574 Wimbledon
1575 Wimbledon
1576 Wimbledon
1577 Wimbledon
1578 Wimbledon
1579 Wimbl

Pandas allows to do better then using python sets.

In [27]:
df['Tournament'].unique()
# see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.unique.html

array(['Brisbane International', 'Qatar Exxon Mobil Open',
       'Maharashtra Open', 'ASB Classic', 'Sydney International',
       'Australian Open', 'Cordoba Open', 'Open Sud de France',
       'Sofia Open', 'Argentina Open', 'New York Open',
       'ABN AMRO World Tennis Tournament', 'Delray Beach Open', 'Open 13',
       'Rio Open', 'Abierto Mexicano', 'Dubai Tennis Championships',
       'Brasil Open', 'BNP Paribas Open', 'Sony Ericsson Open',
       "U.S. Men's Clay Court Championships", 'Grand Prix Hassan II',
       'Monte Carlo Masters', 'Open Banco Sabadell ', 'Hungarian Open',
       'Millennium Estoril Open', 'BMW Open', 'Mutua Madrid Open',
       "Internazionali BNL d'Italia", 'Geneva Open', 'Lyon Open',
       'French Open', 'Rosmalen Grass Court Championships',
       'Mercedes Cup', 'Halle Open', "Queen's Club Championships",
       'Antalya Open', 'Eastbourne International', 'Wimbledon',
       'SkiStar Swedish Open', 'Hall of Fame Championships',
       'Croatia Open

In [28]:
df['Tournament'].value_counts()
# see: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html

French Open                           127
Australian Open                       127
US Open                               127
Wimbledon                             127
Sony Ericsson Open                     95
                                     ... 
Rosmalen Grass Court Championships     27
Antalya Open                           27
Eastbourne International               27
SkiStar Swedish Open                   27
Masters Cup                            15
Name: Tournament, Length: 66, dtype: int64

In [29]:
# Note it returns a Series
for tournament, count in df['Tournament'].value_counts().iteritems():
    print ("During the",tournament, "there were", count, "matches.")

During the French Open there were 127 matches.
During the Australian Open there were 127 matches.
During the US Open there were 127 matches.
During the Wimbledon there were 127 matches.
During the Sony Ericsson Open there were 95 matches.
During the BNP Paribas Open there were 95 matches.
During the Internazionali BNL d'Italia there were 55 matches.
During the Mutua Madrid Open there were 55 matches.
During the Western & Southern Financial Group Masters there were 55 matches.
During the Monte Carlo Masters there were 55 matches.
During the Rogers Masters there were 55 matches.
During the Shanghai Masters there were 55 matches.
During the Winston-Salem Open at Wake Forest University there were 47 matches.
During the BNP Paribas Masters there were 47 matches.
During the Open Banco Sabadell  there were 47 matches.
During the Citi Open there were 47 matches.
During the Qatar Exxon Mobil Open there were 31 matches.
During the Halle Open there were 31 matches.
During the Queen's Club Champio

# Question: Find player with most wins

In [30]:
import pandas as pd

dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'
df = pd.read_excel(dataset_file) 

winners = df['Winner'].value_counts()
most_winner = winners.index[0]
most_winner

'Medvedev D.'

## A focus  on Series

- **Series** is a one-dimensional array-like object containing a sequence of values (of similar types to NumPy types) and an associated array of data labels, called its *index*.

It provides functionalities similar to that of python lists and python dictionaries.

As for Dataframes, Series have an index. The default index is made by numbers from 0-1, otherwise we can specify it, and we can use arbitrary labels as index.


In [31]:
import pandas as pd

pds = pd.Series([4, 7, -5, -3])

print(pds)

0    4
1    7
2   -5
3   -3
dtype: int64


In [32]:
pds = pd.Series( [4, 7, -5, -3], 
                 index=["George", "John", "Paul", "Ringo"])

# Series my have a name and its index may also have a name
pds.name = "The Beatles"
pds.index.name = "Member"

print(pds)

Member
George    4
John      7
Paul     -5
Ringo    -3
Name: The Beatles, dtype: int64


### Indexing and slicing

Series indexing works analogously to Python list if you want to access by position, or similar to python dictionaries if you want to access by index value.

In [33]:
# access by position
print (pds[2])
print ()

# access by index value
print (pds['Paul'])

-5

-5


In [34]:
# Fancy indexing by position
print (pds[ [0,3] ])
print ()

# Fancy indexing by index value
p = pds[  ['John', 'Ringo']  ]
print ( p )
print ()

# note: they return pandas series
print (type(p))

Member
George    4
Ringo    -3
Name: The Beatles, dtype: int64

Member
John     7
Ringo   -3
Name: The Beatles, dtype: int64

<class 'pandas.core.series.Series'>


In [35]:
# We also have Boolean Indexing
print (pds[ [True, False, False, True] ])

Member
George    4
Ringo    -3
Name: The Beatles, dtype: int64


## Slicing, loc and iloc

Pandas tries to understand whether you are using a position or an index value. (What if they are both integers?)

To avoid confusion, one good recommendation is to use `loc` and `iloc`.

#### Example
As in Python, we can use slicing with positions. **Right extreme is NOT included**.

Recommendation: use `iloc`

In [36]:
print ( pds )
print ()

print ( pds[1:3] )
print ()

print ( pds.iloc[1:3] )
print ()

Member
George    4
John      7
Paul     -5
Ringo    -3
Name: The Beatles, dtype: int64

Member
John    7
Paul   -5
Name: The Beatles, dtype: int64

Member
John    7
Paul   -5
Name: The Beatles, dtype: int64



#### Example
But we can also use slicing with labels. **Right extreme is included**. Index must be sorted !

Recommendation: use `loc`

In [37]:
print ( pds['George':'P'] )     # 'P' is not present
print ()

print ( pds['George':'Paul'] )
print ()

print ( pds.loc['George':'Paul'] )


Member
George    4
John      7
Name: The Beatles, dtype: int64

Member
George    4
John      7
Paul     -5
Name: The Beatles, dtype: int64

Member
George    4
John      7
Paul     -5
Name: The Beatles, dtype: int64


As for a dictionary you can check for presence.

In [38]:
print("'John' in pds:", "John" in pds )
print("'Mark' in pds:", "Mark" in pds )

# try
# pds["Mark"] # KeyError exception if not present

'John' in pds: True
'Mark' in pds: False


### Series are mutable

We can change its values with an assignment (also with slicing). 

In [39]:
print(pds)
print()

pds['Paul'] = 42

pds['George':'Paul'] = 5

print(pds)

Member
George    4
John      7
Paul     -5
Ringo    -3
Name: The Beatles, dtype: int64

Member
George    5
John      5
Paul      5
Ringo    -3
Name: The Beatles, dtype: int64


In [40]:
pds['Gennaro'] = 12 # It's a honor be part of this group

print(pds)

Member
George      5
John        5
Paul        5
Ringo      -3
Gennaro    12
Name: The Beatles, dtype: int64


### Filtering + Boolean Indexing

We can filter entries of a Series.

(You have the same in NumPy's Boolean indexing).

#### Example
Here we get only rows with a positive value. 

In [41]:
a = pds > 0
print("pds > 0\n", a)
print()

print("pds[ pds > 0]:\n", 
       pds[ pds > 0 ]  ) # get only positive values of pds 
print("Fab Four are back!")

pds > 0
 Member
George      True
John        True
Paul        True
Ringo      False
Gennaro     True
Name: The Beatles, dtype: bool

pds[ pds > 0]:
 Member
George      5
John        5
Paul        5
Gennaro    12
Name: The Beatles, dtype: int64
Fab Four are back!


In [42]:
# Bye-bye gennaro

psd = pds.drop('Gennaro')

### Operations
We can perform NumPy operations on a Series. 

In [43]:
import numpy as np

print("pds*2:\n", pds*2 )
print()

print("np.exp( pds ):\n", np.exp( pds ) )

pds*2:
 Member
George     10
John       10
Paul       10
Ringo      -6
Gennaro    24
Name: The Beatles, dtype: int64

np.exp( pds ):
 Member
George        148.413159
John          148.413159
Paul          148.413159
Ringo           0.049787
Gennaro    162754.791419
Name: The Beatles, dtype: float64


### Series from a dictionary

Series are so close to a dictionary that you can create one from a dictionary.

In [44]:
sdata = {'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}
pds = pd.Series(sdata)

print(pds)

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64


When passing also an index, this is used to filter matching entries of the dictionary.

In [45]:
states = ['California', 'Ohio', 'Oregon', 'Texas']
pds = pd.Series(sdata, index=states) # California is not in sdata

print(pds) # Welcome missing values

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64


Note that `'Utah'` was not included, and that a special value `NaN` is used for the index `California`.

Missing data is common, (e.g., movies withouth ratings), and usually `NaN` is used to represent them. 

It is possible to use `isnull` to find null values, and to replace them.

In [46]:
obj = pd.Series({'Ohio': 35000, 'Texas': 71000, 
                 'Oregon': 16000, 'Utah': 5000},
               index=['Ohio', 'Texas', 'California'])

print (pd.isnull(obj))
print ()

obj [ pd.isnull(obj) ] = 0.0

print (obj)

Ohio          False
Texas         False
California     True
dtype: bool

Ohio          35000.0
Texas         71000.0
California        0.0
dtype: float64


Alternatively one can use the `fillna` method.

In [47]:
obj = pd.Series({'Ohio': 35000, 'Texas': 71000, 
                 'Oregon': 16000, 'Utah': 5000},
               index=['Ohio', 'Texas', 'California'])

obj2 = obj.fillna(0.0) # returns a new data frame
print (obj)
print ()
print (obj2)

Ohio          35000.0
Texas         71000.0
California        NaN
dtype: float64

Ohio          35000.0
Texas         71000.0
California        0.0
dtype: float64


In [48]:
obj = pd.Series({'Ohio': 35000, 'Texas': 71000, 
                 'Oregon': 16000, 'Utah': 5000},
               index=['Ohio', 'Texas', 'California'])

obj.fillna(0.0, inplace=True) # inplace modification
print (obj)

Ohio          35000.0
Texas         71000.0
California        0.0
dtype: float64


### Allignment by index

A useful Series feature for many applications is that it automatically aligns by index
label in arithmetic operations. 

In [49]:
pds1 = pd.Series({'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000})  # California is missing
  
pds2 = pd.Series({'California': 40000, 'Texas': 555, 'Oregon': 111, 'Utah': 222}) # Ohio is missing

pds1 + pds2

California        NaN
Ohio              NaN
Oregon        16111.0
Texas         71555.0
Utah           5222.0
dtype: float64

In [50]:
pds1 = pd.Series({'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000})  # California is missing
  
pds2 = pd.Series({'California': 40000, 'Texas': 555, 'Oregon': 111, 'Utah': 222}) # Ohio is missing

pds1.add(pds2, fill_value=0)

California    40000.0
Ohio          35000.0
Oregon        16111.0
Texas         71555.0
Utah           5222.0
dtype: float64

## Question: List the player names, and count the number of matches they had


In [51]:
import pandas as pd

dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'
df = pd.read_excel(dataset_file) 
print(df['Winner'].value_counts())
print(df['Loser'].value_counts())
players = df['Winner'].value_counts() + df['Loser'].value_counts()
players.sort_values()

Medvedev D.       58
Djokovic N.       54
Nadal R.          52
Federer R.        51
Tsitsipas S.      50
                  ..
Clarke J.          1
Ymer E.            1
Serdarusic N.      1
King K.            1
Berrettini M.      1
Name: Winner, Length: 204, dtype: int64
Simon G.         28
Sousa J.         28
Paire B.         28
Fritz T.         28
Millman J.       27
                 ..
Aragone JC        1
Griekspoor T.     1
Watanuki Y.       1
Moraing M.        1
Tsitsipas S.      1
Name: Loser, Length: 291, dtype: int64


Celikbilek A.         2.0
Milojevic N.          2.0
Soeda G.              2.0
Griekspoor T.         2.0
Statham J.            2.0
                     ... 
Weintraub A.          NaN
Young D.              NaN
Zapata Miralles B.    NaN
Zayid M.S.            NaN
Zhang Ze.             NaN
Length: 291, dtype: float64

In [52]:
players = df['Winner'].value_counts().add(df['Loser'].value_counts(), fill_value=0)

# Note: you can sort !
players.sort_values(ascending=False)

Medvedev D.       76.0
Tsitsipas S.      74.0
Djokovic N.       63.0
Schwartzman D.    63.0
Goffin D.         62.0
                  ... 
Safwat M.          1.0
Gaio F.            1.0
Roumane R.         1.0
Rola B.            1.0
Krstin P.          1.0
Length: 291, dtype: float64

## Row Selection, Addition, and Deletion

Very similar to column operation , with a different way of specifying rows.

More details at https://pandas.pydata.org/pandas-docs/stable/indexing.html.

#### Example (Selection)
We can access rows by using the index value in the special `loc` attribute.

Note that the result is a dataframe.

In [53]:
df.set_index("Location", inplace=True) # Alternative to df = df.set_index("Location")

df.loc['Brisbane'] # get all the matching rows

ATP              Tournament       Date  Series    Court Surface  \
Location                                                                    
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-03  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-03  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-03  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-04  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-04  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-05  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-05  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-06  ATP250  Outdoor    Hard   

                  Round  Best of         Winner          Loser  ...  Lsets  \
Location                                                        ...          
Brisbane      1st Round        3    Dimitrov G.    Nishioka Y.  ...    0.0   
Brisbane      1st Round        3      Raonic M.      Bedene A.  ...    0.0   
Brisbane      1st Round        3  Kecmanovic M.       Mayer L.  ...    0.0   
Brisbane      1st Round        3     Millman J.    Sandgren T.  ...    1.0   
Brisbane      1st Round        3    Uchiyama Y.     Humbert U.  ...    0.0   
Brisbane      1st Round        3       Kudla D.       Fritz T.  ...    1.0   
Brisbane      1st Round        3      Chardy J.    Struff J.L.  ...    1.0   
Brisbane      1st Round        3      Murray A.   Duckworth J.  ...    0.0   
Brisbane      1st Round        3     Kyrgios N.    Harrison R.  ...    1.0   
Brisbane      1st Round        3    Tsonga J.W.  Kokkinakis T.  ...    0.0   
Brisbane      1st Round        3   De Minaur A.     Popyrin A.  ...    0.0   
Brisbane      1st Round        3    Thompson J.        Bolt A.  ...    0.0   
Brisbane      2nd Round        3    Dimitrov G.     Millman J.  ...    0.0   
Brisbane      2nd Round        3    Uchiyama Y.      Edmund K.  ...    0.0   
Brisbane      2nd Round        3   Nishikori K.       Kudla D.  ...    0.0   
Brisbane      2nd Round        3      Raonic M.  Kecmanovic M.  ...    0.0   
Brisbane      2nd Round        3      Chardy J.     Kyrgios N.  ...    1.0   
Brisbane      2nd Round        3    Medvedev D.      Murray A.  ...    0.0   
Brisbane      2nd Round        3   De Minaur A.    Thompson J.  ...    0.0   
Brisbane      2nd Round        3    Tsonga J.W.      Daniel T.  ...    0.0   
Brisbane  Quarterfinals        3      Chardy J.    Uchiyama 

#### Example (Selection)
We can access rows by using a list of index values.

In [54]:
df.loc[ ['Paris','London'] ] 

ATP   Tournament       Date       Series    Court Surface  \
Location                                                              
Paris      32  French Open 2019-05-26   Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26   Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26   Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26   Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26   Grand Slam  Outdoor    Clay   
...       ...          ...        ...          ...      ...     ...   
London     66  Masters Cup 2019-11-15  Masters Cup   Indoor    Hard   
London     66  Masters Cup 2019-11-15  Masters Cup   Indoor    Hard   
London     66  Masters Cup 2019-11-16  Masters Cup   Indoor    Hard   
London     66  Masters Cup 2019-11-16  Masters Cup   Indoor    Hard   
London     66  Masters Cup 2019-11-17  Masters Cup   Indoor    Hard   

                Round  Best of         Winner          Loser  ...  Lsets  \
Location                                                      ...          
Paris       1st Round        5       Cilic M.    Fabbiano T.  ...    0.0   
Paris       1st Round        5     Popyrin A.     Humbert U.  ...    1.0   
Paris       1st Round        5        Ruud C.      Gulbis E.  ...    0.0   
Paris       1st Round        5  Berrettini M.     Andujar P.  ...    1.0   
Paris       1st Round        5    Dimitrov G.  Tipsarevic J.  ...    2.0   
...               ...      ...            ...            ...  ...    ...   
London    Round Robin        3      Nadal R.   Tsitsipas S.   ...    1.0   
London    Round Robin        3     Zverev A.    Medvedev D.   ...    0.0   
London     Semifinals        3  Tsitsipas S.     Federer R.   ...    0.0   
London     Semifinals        3      Thiem D.      Zverev A.   ...    0.0   
London      The Final        3  Tsitsipas S.       Thiem D.   ...    1.0   

            Comment  B365W  B365L   PSW    PSL  MaxW   MaxL  AvgW  AvgL  
Location                                                                 
Paris     Completed   1.08   8.00  1.08  10.62  1.11  10.62  1.08  8.01  
Paris     Completed   2.00   1.80  2.12   1.81  2.18   1.85  2.04  1.78  
Paris     Completed   1.28   3.75  1.28   4.15  1.31   4.15  1.27  3.81  
Paris     Completed   1.20   4.50  1.19   5.42  1.22   5.42  1.18  4.81  
Paris     Completed   1.16   5.00  1.20   5.21  1.20   5.50  1.18  5.00  
...             ...    ...    ...   ...    ...   ...    ...   ...   ...  
London    Completed   1.44   2.75  1.39   3.26  1.48   3.30  1.41  2.93  
London    Completed   1.90   1.90  2.14   1.79  2.24   2.06  1.92  1.90  
London    Completed   3.50   1.30  3.75   1.33  3.75   1.40  3.39  1.33  
London    Completed   1.80   2.00  1.84   2.10  1.87   2.20  1.78  2.06  
London    Completed   2.00   1.80  2.00   1.93  2.05   1.93  1.96  1.86  

[316 rows x 35 columns]

#### Example (Selection)
Rows can be selected by using integer location with the special `iloc` attribute.

In [55]:
df.iloc[2]

ATP                                1
Tournament    Brisbane International
Date             2018-12-31 00:00:00
Series                        ATP250
Court                        Outdoor
Surface                         Hard
Round                      1st Round
Best of                            3
Winner                 Kecmanovic M.
Loser                       Mayer L.
WRank                          131.0
LRank                           56.0
WPts                           433.0
LPts                           895.0
W1                               6.0
L1                               3.0
W2                               6.0
L2                               1.0
W3                               NaN
L3                               NaN
W4                               NaN
L4                               NaN
W5                               NaN
L5                               NaN
Wsets                            2.0
Lsets                            0.0
Comment                    Completed
B

#### Example (Selection)
Slicing over the rows with `iloc`.

In [56]:
df.iloc[1:10]

ATP              Tournament       Date  Series    Court Surface  \
Location                                                                    
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   

              Round  Best of         Winner          Loser  ...  Lsets  \
Location                                                    ...          
Brisbane  1st Round        3      Raonic M.      Bedene A.  ...    0.0   
Brisbane  1st Round        3  Kecmanovic M.       Mayer L.  ...    0.0   
Brisbane  1st Round        3     Millman J.    Sandgren T.  ...    1.0   
Brisbane  1st Round        3    Uchiyama Y.     Humbert U.  ...    0.0   
Brisbane  1st Round        3       Kudla D.       Fritz T.  ...    1.0   
Brisbane  1st Round        3      Chardy J.    Struff J.L.  ...    1.0   
Brisbane  1st Round        3      Murray A.   Duckworth J.  ...    0.0   
Brisbane  1st Round        3     Kyrgios N.    Harrison R.  ...    1.0   
Brisbane  1st Round        3    Tsonga J.W.  Kokkinakis T.  ...    0.0   

            Comment  B365W  B365L   PSW   PSL  MaxW  MaxL  AvgW  AvgL  
Location                                                               
Brisbane  Completed   1.18   4.50  1.23  4.68  1.27  4.84  1.22  4.26  
Brisbane  Completed   1.57   2.25  1.67  2.32  1.71  2.40  1.63  2.28  
Brisbane  Completed   1.40   2.75  1.41  3.13  1.45  3.20  1.40  2.95  
Brisbane  Completed   2.62   1.44  2.73  1.51  3.26  1.53  2.69  1.47  
Brisbane  Completed   2.62   1.44  2.80  1.49  2.85  1.55  2.70  1.47  
Brisbane  Completed   2.10   1.66  2.23  1.72  2.26  1.74  2.19  1.68  
Brisbane  Completed   1.28   3.50  1.38  3.29  1.39  3.60  1.34  3.26  
Brisbane  Completed   1.40   2.75  1.47  2.87  1.50  3.16  1.44  2.80  
Brisbane  Completed   2.25   1.57  2.27  1.70  2.42  1.71  2.27  1.64  

[9 rows x 35 columns]

#### Example (Addition)
We can add a new row by using `loc` attribute.

Pandas also has an `append` function (see https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.append.html).

In [57]:
# Make a smaller copy
small_df = pd.DataFrame( df[ ['Surface', 'Winner', 'Loser'] ] ) 
small_df = small_df.loc[['London', 'Paris']]

small_df.loc['Mestre'] = ['Clay', 'Claudio L.', 'Nadal R.']
small_df.loc['Mestre'] = ['Clay', 'Claudio L.', 'Federer R.'] # This is a replace

small_df

Surface            Winner          Loser
Location                                         
London     Grass     Tipsarevic J.    Nishioka Y.
London     Grass       Anderson K.   Herbert P.H.
London     Grass          Lopez F.       Giron M.
London     Grass  Bautista Agut R.   Gojowczyk P.
London     Grass         Opelka R.     Stebe C.M.
...          ...               ...            ...
Paris       Hard          Nadal R.    Tsonga J.W.
Paris       Hard       Djokovic N.    Dimitrov G.
Paris       Hard     Shapovalov D.       Nadal R.
Paris       Hard       Djokovic N.  Shapovalov D.
Mestre      Clay        Claudio L.     Federer R.

[317 rows x 3 columns]

#### Example (Deletion)
We can remove a row (or more) with `drop()` function.

In [58]:
small_df.drop("Mestre", inplace=True) # axis=0 is the default
small_df

Surface            Winner          Loser
Location                                         
London     Grass     Tipsarevic J.    Nishioka Y.
London     Grass       Anderson K.   Herbert P.H.
London     Grass          Lopez F.       Giron M.
London     Grass  Bautista Agut R.   Gojowczyk P.
London     Grass         Opelka R.     Stebe C.M.
...          ...               ...            ...
Paris       Hard     Shapovalov D.     Monfils G.
Paris       Hard          Nadal R.    Tsonga J.W.
Paris       Hard       Djokovic N.    Dimitrov G.
Paris       Hard     Shapovalov D.       Nadal R.
Paris       Hard       Djokovic N.  Shapovalov D.

[316 rows x 3 columns]

In [59]:
# Make a smaller copy
small_df = pd.DataFrame( df[ ['Surface', 'Winner', 'Loser'] ] ) 
small_df = small_df.loc[['London', 'Paris']]

small_df.loc['Mestre'] = ['Clay', 'Claudio L.', 'Nadal R.']
small_df.loc['Mestre'] = ['Clay', 'Claudio L.', 'Federer R.'] # This is a replace

# Remove all matching rows
small_df.drop(["London", "Paris"], inplace=True)
small_df

Surface      Winner       Loser
Location                                
Mestre      Clay  Claudio L.  Federer R.

### Creating a DataFrame

There are many ways to construct a DataFrame, though one of the most common is from a dictionary of equal-length lists (or NumPy arrays).

In [60]:
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003],
        'population': [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}

df = pd.DataFrame(data) 

df.index.name = "Progressive"   # Set index name
df.columns.name = 'Attributes'  # Set columns name

df # Default index is 0...N-1 

Attributes    state  year  population
Progressive                          
0              Ohio  2000         1.5
1              Ohio  2001         1.7
2              Ohio  2002         3.6
3            Nevada  2001         2.4
4            Nevada  2002         2.9
5            Nevada  2003         3.2

In [61]:
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003],
        'population': [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}

df = pd.DataFrame(data,
                 index = ['xxx1','xxx2','xxx3','xxx4','xxx5','xxx6']) 

df.index.name   = "Custom ID"   # Set index name
df.columns.name = 'Attributes'  # Set columns name

df

Attributes   state  year  population
Custom ID                           
xxx1          Ohio  2000         1.5
xxx2          Ohio  2001         1.7
xxx3          Ohio  2002         3.6
xxx4        Nevada  2001         2.4
xxx5        Nevada  2002         2.9
xxx6        Nevada  2003         3.2

## From pandas to numpy

If you prefer working with numpy

In [62]:
m = df.values
print (type(m))
print ()

print (m)

<class 'numpy.ndarray'>

[['Ohio' 2000 1.5]
 ['Ohio' 2001 1.7]
 ['Ohio' 2002 3.6]
 ['Nevada' 2001 2.4]
 ['Nevada' 2002 2.9]
 ['Nevada' 2003 3.2]]


# Question: Find the most unexpected win by Nadal with largest odds

In [63]:
import pandas as pd

dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'

df = pd.read_excel(dataset_file) 

In [64]:
nadal_winner = df[ df['Winner']=='Nadal R.' ]
nadal_winner = nadal_winner[ ['Winner', 'B365W'] ]
nadal_winner.columns = ['Player','bet'] # see also rename function
nadal_winner.head()

Player   bet
149  Nadal R.  1.05
218  Nadal R.  1.06
241  Nadal R.  1.10
252  Nadal R.  1.14
260  Nadal R.  1.04

In [65]:
nadal_winner.sort_values(by='bet',ascending=False, inplace=True)
nadal_winner.head()

Player   bet
1150  Nadal R.  1.44
1950  Nadal R.  1.40
1595  Nadal R.  1.25
2586  Nadal R.  1.25
1148  Nadal R.  1.25

In [66]:
match_id = nadal_winner.index[1]
print (df.loc[match_id])

ATP                            49
Location                 Montreal
Tournament         Rogers Masters
Date          2019-08-11 00:00:00
Series               Masters 1000
Court                     Outdoor
Surface                      Hard
Round                   The Final
Best of                         3
Winner                   Nadal R.
Loser                 Medvedev D.
WRank                         2.0
LRank                         9.0
WPts                       7945.0
LPts                       2745.0
W1                            6.0
L1                            3.0
W2                            6.0
L2                            0.0
W3                            NaN
L3                            NaN
W4                            NaN
L4                            NaN
W5                            NaN
L5                            NaN
Wsets                         2.0
Lsets                         0.0
Comment                 Completed
B365W                         1.4
B365L         

# Question: how many times the player with the best ranking won the match?


In [67]:
import pandas as pd

dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'

df = pd.read_excel(dataset_file) 

In [68]:
df.columns

Index(['ATP', 'Location', 'Tournament', 'Date', 'Series', 'Court', 'Surface',
       'Round', 'Best of', 'Winner', 'Loser', 'WRank', 'LRank', 'WPts', 'LPts',
       'W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'W4', 'L4', 'W5', 'L5', 'Wsets',
       'Lsets', 'Comment', 'B365W', 'B365L', 'PSW', 'PSL', 'MaxW', 'MaxL',
       'AvgW', 'AvgL'],
      dtype='object')

In [69]:
w_gt_l = df['WRank']>df['LRank']

w_gt_l.head()

0    False
1    False
2     True
3    False
4     True
dtype: bool

In [70]:
w_gt_l = w_gt_l.astype(int)

w_gt_l.head()

0    0
1    0
2    1
3    0
4    1
dtype: int64

In [71]:
total_wins = w_gt_l.sum()

total_wins

1004

In [72]:
total_matches, _ = df.shape

total_matches

2610

In [73]:
print ("The success rate of the best ranked player is", round(100.0*total_wins/total_matches,2), "%" )

The success rate of the best ranked player is 38.47 %


# Question: how much would you gain or lose by always betting 10€ on the best ranked player?

In [74]:
import pandas as pd

dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'

df = pd.read_excel(dataset_file) 

In [75]:
w_gt_l = df['WRank']>df['LRank']

In [76]:
# decompose the following to understand each singles setps
gains = ( (df['B365W'][w_gt_l]-1.0) * 5.0).sum()

In [77]:
losses = (~w_gt_l).sum() * 5.0

In [78]:
total = gains - losses

print ("If always betting on the best ranked, the profit would be", total)

If always betting on the best ranked, the profit would be -943.7999999999993


## Additional useful manipulation

Suppose, for some reason, you want to invert name/surname order.

That is, suppose you want to apply the same &custom* function to every element of a dataframe/series.

In [79]:
def my_fun (x):
    tokens = x.split()
    tokens = tokens[::-1]
    new_x  = ' '.join(tokens)
    return new_x

# note the re-assiggnment
df['Winner'] = df['Winner'].map(my_fun)

df['Winner']

0         G. Dimitrov
1           M. Raonic
2       M. Kecmanovic
3          J. Millman
4         Y. Uchiyama
            ...      
2605         R. Nadal
2606        A. Zverev
2607     S. Tsitsipas
2608         D. Thiem
2609     S. Tsitsipas
Name: Winner, Length: 2610, dtype: object

## Question: What is the surface with longest matches on average (more games)?

In [80]:
df.columns

Index(['ATP', 'Location', 'Tournament', 'Date', 'Series', 'Court', 'Surface',
       'Round', 'Best of', 'Winner', 'Loser', 'WRank', 'LRank', 'WPts', 'LPts',
       'W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'W4', 'L4', 'W5', 'L5', 'Wsets',
       'Lsets', 'Comment', 'B365W', 'B365L', 'PSW', 'PSL', 'MaxW', 'MaxL',
       'AvgW', 'AvgL'],
      dtype='object')

In [81]:
sub_df = df[ ['W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'W4', 'L4', 'W5', 'L5'] ]
sub_df.head()

W1   L1   W2   L2   W3   L3  W4  L4  W5  L5
0  6.0  3.0  6.0  4.0  NaN  NaN NaN NaN NaN NaN
1  6.0  0.0  6.0  3.0  NaN  NaN NaN NaN NaN NaN
2  6.0  3.0  6.0  1.0  NaN  NaN NaN NaN NaN NaN
3  7.0  6.0  6.0  7.0  6.0  0.0 NaN NaN NaN NaN
4  6.0  4.0  7.0  6.0  NaN  NaN NaN NaN NaN NaN

In [82]:
sub_df = sub_df.fillna(0.0)
sub_df.head()

W1   L1   W2   L2   W3   L3   W4   L4   W5   L5
0  6.0  3.0  6.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0
1  6.0  0.0  6.0  3.0  0.0  0.0  0.0  0.0  0.0  0.0
2  6.0  3.0  6.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
3  7.0  6.0  6.0  7.0  6.0  0.0  0.0  0.0  0.0  0.0
4  6.0  4.0  7.0  6.0  0.0  0.0  0.0  0.0  0.0  0.0

In [83]:
sub_df.sum()

W1    15085.0
L1    10879.0
W2    15029.0
L2    10399.0
W3     7562.0
L3     4636.0
W4     1538.0
L4     1083.0
W5      619.0
L5      360.0
dtype: float64

In [84]:
sub_df.sum(axis=1)

0       19.0
1       15.0
2       16.0
3       32.0
4       23.0
        ... 
2605    35.0
2606    23.0
2607    19.0
2608    21.0
2609    34.0
Length: 2610, dtype: float64

In [85]:
df['games'] = sub_df.sum(axis=1)

In [86]:
df[ ['Surface','games'] ].groupby('Surface').mean()

games
Surface           
Clay     24.969349
Grass    28.904321
Hard     25.465070

## Pivot Tables

Similar to groupby, also columsn are grouped.

 - See https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.pivot_table.html
 
The function `pivot_table` has the following main parameters.

| Param name | Description|
|---|:--|
| values  | The column to be aggregated |
| index   | Column names used to create rows of the Pivot Table |
| columns | Column names used to create cols of the Pivot Table  |
| aggfunc | Aggregation function (e.g., `np.sum`)


# Question: Find the most successful player by surface 

In [87]:
import pandas as pd
dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'
df = pd.read_excel(dataset_file) 


In [88]:
df.head()

ATP  Location              Tournament       Date  Series    Court Surface  \
0    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
1    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
2    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
3    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
4    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   

       Round  Best of         Winner  ... Lsets    Comment  B365W  B365L  \
0  1st Round        3    Dimitrov G.  ...   0.0  Completed   1.36   3.00   
1  1st Round        3      Raonic M.  ...   0.0  Completed   1.18   4.50   
2  1st Round        3  Kecmanovic M.  ...   0.0  Completed   1.57   2.25   
3  1st Round        3     Millman J.  ...   1.0  Completed   1.40   2.75   
4  1st Round        3    Uchiyama Y.  ...   0.0  Completed   2.62   1.44   

    PSW   PSL  MaxW  MaxL  AvgW  AvgL  
0  1.36  3.37  1.42  3.60  1.35  3.18  
1  1.23  4.68  1.27  4.84  1.22  4.26  
2  1.67  2.32  1.71  2.40  1.63  2.28  
3  1.41  3.13  1.45  3.20  1.40  2.95  
4  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 36 columns]

In [89]:
import pandas as pd
dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'
df = pd.read_excel(dataset_file) 


df.pivot_table(values="ATP", # irrelevant
               index="Winner",
               columns="Surface",
               aggfunc=len)

Surface             Clay  Grass  Hard
Winner                               
Albot R.             7.0    1.0  23.0
Anderson K.          NaN    3.0   8.0
Andreozzi G.         2.0    1.0   3.0
Andujar P.           9.0    NaN   6.0
Auger-Aliassime F.  12.0    9.0  12.0
...                  ...    ...   ...
Ymer M.              2.0    NaN   1.0
Zhang Zh.            NaN    NaN   2.0
Zverev A.           16.0    2.0  21.0
Zverev A.            NaN    NaN   2.0
Zverev M.            NaN    2.0   1.0

[204 rows x 3 columns]

In [90]:
df.pivot_table(values="ATP", # irrelevant
               index="Winner",
               columns="Surface",
               aggfunc=len,
               fill_value=0.0 )

Surface             Clay  Grass  Hard
Winner                               
Albot R.               7      1    23
Anderson K.            0      3     8
Andreozzi G.           2      1     3
Andujar P.             9      0     6
Auger-Aliassime F.    12      9    12
...                  ...    ...   ...
Ymer M.                2      0     1
Zhang Zh.              0      0     2
Zverev A.             16      2    21
Zverev A.              0      0     2
Zverev M.              0      2     1

[204 rows x 3 columns]

In [91]:
wins = df.pivot_table(values="ATP", # irrelevant
               index="Winner",
               columns="Surface",
               aggfunc=len,
               fill_value=0.0 )
wins.head()

Surface             Clay  Grass  Hard
Winner                               
Albot R.               7      1    23
Anderson K.            0      3     8
Andreozzi G.           2      1     3
Andujar P.             9      0     6
Auger-Aliassime F.    12      9    12

In [92]:
wins.columns

Index(['Clay', 'Grass', 'Hard'], dtype='object', name='Surface')

In [93]:
for surface in wins.columns:
    print (wins[surface].sort_values(ascending=False).head(1))
    print ()

Winner
Thiem D.    23
Name: Clay, dtype: int64

Winner
Berrettini M.    11
Name: Grass, dtype: int64

Winner
Medvedev D.    45
Name: Hard, dtype: int64



# Question: Find the player with the best success rate  by surface

In [94]:
import pandas as pd
dataset_file = 'http://tennis-data.co.uk/2019/2019.xlsx'
df = pd.read_excel(dataset_file) 

In [95]:
wins = df.pivot_table(values="ATP", # irrelevant
               index="Winner",
               columns="Surface",
               aggfunc=len,
               fill_value=0.0 )
wins.head()

Surface             Clay  Grass  Hard
Winner                               
Albot R.               7      1    23
Anderson K.            0      3     8
Andreozzi G.           2      1     3
Andujar P.             9      0     6
Auger-Aliassime F.    12      9    12

In [96]:
losses = df.pivot_table(values="ATP", # irrelevant
               index="Loser",
               columns="Surface",
               aggfunc=len,
               fill_value=0.0 )
losses.head()

Surface       Clay  Grass  Hard
Loser                          
Albot R.         6      3    17
Altmaier D.      1      0     0
Anderson K.      0      2     2
Andreev A.       0      0     1
Andreozzi G.     8      2     4

In [97]:
total = wins + losses
total.head()

Surface       Clay  Grass  Hard
Albot R.      13.0    4.0  40.0
Altmaier D.    NaN    NaN   NaN
Anderson K.    0.0    5.0  10.0
Andreev A.     NaN    NaN   NaN
Andreozzi G.  10.0    3.0   7.0

In [98]:
total = wins.add(losses, fill_value=0)
total.head()

Surface       Clay  Grass  Hard
Albot R.      13.0    4.0  40.0
Altmaier D.    1.0    0.0   0.0
Anderson K.    0.0    5.0  10.0
Andreev A.     0.0    0.0   1.0
Andreozzi G.  10.0    3.0   7.0

In [99]:
wins.loc['Nadal R.']

Surface
Clay     21
Grass     5
Hard     26
Name: Nadal R., dtype: int64

In [100]:
losses.loc['Nadal R.']

Surface
Clay     3
Grass    1
Hard     4
Name: Nadal R., dtype: int64

In [101]:
total.loc['Nadal R.']

Surface
Clay     24.0
Grass     6.0
Hard     30.0
Name: Nadal R., dtype: float64

In [102]:
success_rate = wins / total

success_rate.head()

Surface           Clay     Grass      Hard
Albot R.      0.538462  0.250000  0.575000
Altmaier D.        NaN       NaN       NaN
Anderson K.        NaN  0.600000  0.800000
Andreev A.         NaN       NaN       NaN
Andreozzi G.  0.200000  0.333333  0.428571

In [103]:
success_rate = wins / total
success_rate.fillna(0.0, inplace=True)
success_rate.head()

Surface           Clay     Grass      Hard
Albot R.      0.538462  0.250000  0.575000
Altmaier D.   0.000000  0.000000  0.000000
Anderson K.   0.000000  0.600000  0.800000
Andreev A.    0.000000  0.000000  0.000000
Andreozzi G.  0.200000  0.333333  0.428571

In [104]:
success_rate.loc['Nadal R.']

Surface
Clay     0.875000
Grass    0.833333
Hard     0.866667
Name: Nadal R., dtype: float64

In [105]:
success_rate.loc['Thiem D.']

Surface
Clay     0.766667
Grass    0.000000
Hard     0.758621
Name: Thiem D., dtype: float64

In [106]:
for surface in success_rate:  # note: you do not need to use .columns
    print (success_rate[surface].sort_values(ascending=False).head(1))
    print ()

Nadal R.    0.875
Name: Clay, dtype: float64

Djokovic N.    1.0
Name: Grass, dtype: float64

Nadal R.    0.866667
Name: Hard, dtype: float64



## Other Useful stuff



### DataFrame Joins

#### Many-to-One join
Consider the following two DataFrames.

In [107]:
df1 = pd.DataFrame({'key': ['b', 'b', 'a', 'c', 'a', 'a', 'b'], 
                    'data1': range(7)})

df2 = pd.DataFrame({'key': ['a', 'b', 'd'], 
                    'data2': range(3)})

print(df1)
print(df2)

  key  data1
0   b      0
1   b      1
2   a      2
3   c      3
4   a      4
5   a      5
6   b      6
  key  data2
0   a      0
1   b      1
2   d      2


The data in df1 has multiple rows labeled a and b, whereas df2 has only one row for each value in the key column. 

The operation 

```
pd.merge(df1, df2, on='key')
```

will produce a merged DataFrame with key the column to join on. 


In [108]:
pd.merge(df1, df2, on='key')

key  data1  data2
0   b      0      1
1   b      1      1
2   b      6      1
3   a      2      0
4   a      4      0
5   a      5      0

Notice that 
- data2 of df2 is replicated everywhere key 'a' and 'b' occur in df1
- rows with key 'c' in df1 and 'd' in df2 are not present in df

The latter behavior is called **inner** join: the keys in the result are the intersection, or the common set found in both tables.

Other possible approaches are: 
- **left**: keys of left DataFrame are kept
- **right**: keys of right DataFrame are kept
- **outer**: keys of both DataFrames are kept

and can be chosen by setting parameter *how*

#### left

In [109]:
pd.merge(df1, df2, on='key', how='left')

key  data1  data2
0   b      0    1.0
1   b      1    1.0
2   a      2    0.0
3   c      3    NaN
4   a      4    0.0
5   a      5    0.0
6   b      6    1.0

#### right

In [110]:
pd.merge(df1, df2, on='key', how='right')

key  data1  data2
0   a    2.0      0
1   a    4.0      0
2   a    5.0      0
3   b    0.0      1
4   b    1.0      1
5   b    6.0      1
6   d    NaN      2

#### outer

In [111]:
pd.merge(df1, df2, on='key', how='outer')

key  data1  data2
0   b    0.0    1.0
1   b    1.0    1.0
2   b    6.0    1.0
3   a    2.0    0.0
4   a    4.0    0.0
5   a    5.0    0.0
6   c    3.0    NaN
7   d    NaN    2.0

#### Many-to-Many join
Many-to-many merges happens when the same key has more than one occurrence in both DataFrames.

Many-to-many joins form the Cartesian product of the rows having the same key. 

See the following example:

In [112]:
df1 = pd.DataFrame({'key': ['b', 'b', 'a',  'a', 'b'], 
                    'data1': range(5)})
                    
df2 = pd.DataFrame({'key': ['a', 'b', 'a', 'b', 'd'],
                    'data2': range(5)})

print (df1)
print()
print (df2)
print ()

pd.merge(df1, df2, on='key', how='left')

  key  data1
0   b      0
1   b      1
2   a      2
3   a      3
4   b      4

  key  data2
0   a      0
1   b      1
2   a      2
3   b      3
4   d      4



key  data1  data2
0   b      0      1
1   b      0      3
2   b      1      1
3   b      1      3
4   a      2      0
5   a      2      2
6   a      3      0
7   a      3      2
8   b      4      1
9   b      4      3

Note that, since there were three 'b' rows in df1 and two in df2, there are six 'b' rows in the result.

The merge can be performed also respect to more than one variable. To determine which key combinations will appear in the result depending on the choice of merge method, **think of the multiple keys as forming an array of tuples** to be used as a single join key.

# References

 - **Python for Data Analysis**. O'Reilly. Wes McKinney.
   - Section 5.1 (Indexing, Selection, and Filtering)
   - Section 5.2 (Arithmetic and Data Alignment Function Application and Mapping Sorting and Ranking)
   - Section 5.3 (Unique Values, Value Counts, and Membership)
   - Section 10.4 (Pivot Tables)
   - Section 14.5
   - https://github.com/wesm/pydata-book
